rust-bins/src/bin/keygen.rs

```diff
@@ -301,6 +472,228 @@ pub fn generate_ed_sk(
+        if bip39_map.contains_key(&*word) {
+            return Ok(word);
+        } else {
+            println!("The word you have entered is not in the BIP39 word list. Please try again.");
+        }
+    }
+}
+
+/// Read vector of words from file.
+pub fn read_words_from_file(
+    path: PathBuf,
+    verify_bip39: bool,
+    bip39_map: &HashMap<&str, usize>,
+) -> Result<Vec<String>, String> {
+    let word_string = succeed_or_die!(
+        fs::read_to_string(path),
+        e => "Could not read input from provided file because {}"
+    );
+
+    let word_list: Vec<String> = word_string.split_whitespace().map(str::to_owned).collect();
+
+    // verify whether input_words is a valid BIP39 sentence if check is enabled
+    if verify_bip39 && !verify_bib39(&word_list, &bip39_map) {
+        return Err("The input does not constitute a valid BIP39 sentence.".to_string());
+    }
+
+    Ok(word_list)
+}
+
+/// Ask user to input num words.
+/// If verify_bip39 is true, output is verified to be valid BIP39 sentence.
+pub fn read_words_from_terminal(
+    num: u8,
+    verify_bip39: bool,
+    bip39_map: &HashMap<&str, usize>,
+) -> Result<Vec<String>, String> {
+    // Ensure that input length is in allowed set if verification is enabled.
+    if verify_bip39 {
+        match num {
+            12 | 15 | 18 | 21 | 24 => (),
+            _ => {
+                return Err(
+                    "The input length for a valid BIP39 sentence must be in {12, 15, 18, 21, 24}."
+                        .to_string(),
+                )
+            }
+        };
+    }
+
+    let mut word_list = Vec::<String>::new();
+    for i in 1..=num {
+        // read the ith word from stdin
+        let word = match verify_bip39 {
+            false => succeed_or_die!(
+                read_word(i),
+                e => "Could not read input from user because {}"
+            ),
+            true => succeed_or_die!(
+                read_bip39_word(i, &bip39_map),
+                e => "Could not read input from user because {}"
+            ),
+        };
+        word_list.push(word);
+    }
+
+    // verify whether input_words is a valid BIP39 sentence if check is enabled
+    if verify_bip39 && !verify_bib39(&word_list, &bip39_map) {
+        return Err("The input does not constitute a valid BIP39 sentence.".to_string());
+    }
+
+    Ok(word_list)
+}
+
+/// Verify whether the given vector of words constitutes a valid BIP39 sentence.
+pub fn verify_bib39(word_vec: &[String], bip_word_map: &HashMap<&str, usize>) -> bool {
+    // check that word_vec contains allowed number of words
+    match word_vec.len() {
+        12 | 15 | 18 | 21 | 24 => (),
+        _ => return false,
+    };
+
+    // convert word vector to bits
+    let mut bit_vec = BitVec::<Msb0, u8>::new();
+    for word in word_vec {
+        match bip_word_map.get(&**word) {
+            Some(idx) => {
+                let word_bits = BitVec::<Msb0, u16>::from_slice(&[*idx as u16]).unwrap();
```

Comment:
Perhaps document why the unwrap is safe (or use .except with a meaningful string).